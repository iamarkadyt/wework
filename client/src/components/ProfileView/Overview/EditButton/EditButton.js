import React from 'react'
import Field from '../../../Field/Field'
import { func } from 'prop-types'
import {
    FaPencilAlt as IcoEdit
} from 'react-icons/fa'

const editButton = ({
  navTo,
  quitEntryDeletingMode
}) => (
  <Field
    type="button"
    containerStyle={{ margin: 0 }}
    onClick={() => {
      quitEntryDeletingMode()
      navTo('update-profile')
    }}>
    <span style={{ fontSize: '.8rem' }}><IcoEdit /></span>
    &nbsp;
    Edit Profile
  </Field>
)

editButton = {
  navTo: func.isRequired,
  quitEntryDeletingMode: func.isRequired
}

export default editButton
