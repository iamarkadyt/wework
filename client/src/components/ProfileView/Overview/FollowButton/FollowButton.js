import React from 'react'
import Field from '../../../Field/Field'
import { func, string } from 'prop-types'
import {
    FaUserCheck as IcoFollow
} from 'react-icons/fa'

const followButton = ({
  followAPerson,
  profileOwnerId
}) => (
  <Field
    type="button"
    containerStyle={{ margin: 0 }}
    onClick={() => followAPerson(profileOwnerId)}>
    <span style={{ fontSize: '.8rem' }}>
        <IcoFollow />
    </span>
    &nbsp;
    Follow
  </Field>
)

followButton = {
  followAPerson: func.isRequired,
  profileOwnerId: string.isRequired
}

export default followButton
