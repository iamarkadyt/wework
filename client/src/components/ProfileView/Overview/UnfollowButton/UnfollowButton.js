import React from 'react'
import Field from '../../../Field/Field'
import { func, string } from 'prop-types'
import {
    FaUserMinus as IcoUnfollow
} from 'react-icons/fa'

const unfollowButton = ({
  unfollowAPerson,
  profileOwnerId
}) => (
  <Field
    type="button"
    containerStyle={{ margin: 0 }}
    onClick={() => unfollowAPerson(profileOwnerId)}>
    <span style={{ fontSize: '.8rem' }}>
        <IcoUnfollow />
    </span>
    &nbsp;
    Unfollow
  </Field>
)

unfollowButton.propTypes = {
  unfollowAPerson: func.isRequired,
  profileOwnerId: string.isRequired
}

export default unfollowButton
