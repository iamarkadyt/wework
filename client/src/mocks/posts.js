export const mockPost = {
    "_id": "c717a50ac29bacb2ee",
    "user": {
        "_id": "some_alphanumeric_gibberish",
        "name": "Dwayne Johnson",
        "avatar": ""
    },
    "text": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
    "comments": [
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        },
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        },
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        }
    ],
    "likes": [
        {
            "_id": "19825918265102965",
            "user": "aoeu9a8o10s392h0"
        },
        {
            "_id": "19825918265102966",
            "user": "aoeu9a8o10s392hx0"
        },
        {
            "_id": "19825918265102967",
            "user": "aoeu9a8o10s392120"
        }
    ],
    "date": "1535133024"
}

export default new Array(10).fill(mockPost, 0, 10)
