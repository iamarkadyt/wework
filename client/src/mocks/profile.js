export default {
    "skills": [
        "Assembler",
        "Go",
        "C#",
        "C++"
    ],
    "social": [],
    "date": "2018-08-23T20:25:12.263Z",
    "_id": "5b7b42ba57a5806b896e8f0d",
    "user": {
        "_id": "5b75f023f1957052e52106f5",
        "name": "John Doe"
    },
    "__v": 0,
    "handle": "techguy_handle",
    "status": "Actively applying",
    "title": "Senior Developer",
    "company": "Itron",
    "experience": [
        {
            "current": false,
            "_id": "5b7ca6cf63353349cf03756c",
            "title": "SWE",
            "company": "Revature",
            "from": "2009-03-31T23:00:00.000Z",
            "to": "2025-08-01T01:13:50.111Z"
        }
    ],
    "education": [
        {
            "current": true,
            "_id": "5b7cab633345cc4c066c0540",
            "school": "AC Bootcamp",
            "degree": "Nano CS Degree",
            "from": "2015-08-02T18:33:20.000Z"
        }
    ],
    "bio": "My name is [techguyinfo]. I am a senior developer currently working at Microsoft. I specialize in AI and Cloud computing.",
    "githubusername": "arkadyt",
    "location": "Seattle, WA",
    "website": "techguy.info"
}