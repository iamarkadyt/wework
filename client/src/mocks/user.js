export default {
  id: "98ac61abed72ee8d12"
}
