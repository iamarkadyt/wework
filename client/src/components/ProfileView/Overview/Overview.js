import React from 'react'
import Name from './Name/Name'
import Links from './Links/Links'
import Location from './Location/Location'
import EditButton from './EditButton/EditButton'
import FollowButton from './FollowButton/FollowButton'
import UnfollowButton from './UnfollowButton/UnfollowButton'

const overview = ({
  profile: {
    user: {
      _id: profileOwnerId,
      avatar
    },
    status
  },
  profile,
  profileBelongsToAuthedUser,
  isFollowingProfileOwner,
  baseUrl,
  history,
  followAPerson,
  unfollowAPerson,
  quitEntryDeletingMode
}) => {
  const navTo = where => history.push(`${baseUrl}/${where}`)

  let buttons
  if (profileBelongsToAuthedUser) {
    buttons = (
      <EditButton 
        quitEntryDeletingMode={quitEntryDeletingMode}
        navTo={navTo} />
    )
  } else {
    if (isFollowingProfileOwner) {
      buttons = (
        <UnfollowButton 
          unfollowAPerson={unfollowAPerson}
          quitEntryDeletingMode={quitEntryDeletingMode}
          profileOwnerId={profileOwnerId} />
      )
    } else {
      buttons = (
        <FollowButton 
          followAPerson={followAPerson}
          quitEntryDeletingMode={quitEntryDeletingMode}
          profileOwnerId={profileOwnerId} />
      )
    }
  }
  
  return (
    <section className="intro">
      <img src={avatar} alt="" />

      <div className="info">
        <Name profile={profile} />
        <Location profile={profile} />
        <p>{status}</p>
        <Links profile={profile} />
      </div>

      <div className="info-buttons">
        {buttons}
      </div>
    </section>
  )
}

export default overview
