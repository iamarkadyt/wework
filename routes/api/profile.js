const express = require('express')
const router = express.Router()
const mongoose = require('mongoose')
const passport = require('passport')
const Profile = require('../../models/Profile')
const User = require('../../models/User')

// @route   GET api/profile/test
// @desc    Tests profile route.
// @access  Public
router.get('/test', (req, res) => {
    res.json({ message: 'OK' })
})


// @route   GET api/profile
// @desc    Get current users profile
// @access  Private
router.get('/', passport.authenticate('jwt', { session: false }), (req, res) => {
    const errors = {}

    Profile.findOne({ user: req.user.id })
        .then(profile => {
            if (!profile) {
                errors.noprofile = 'There is no profile for this user'
                res.status(404).json({ errors })
            }

            res.json(profile)
        })
        .catch(err => res.status(404).json(err))
})

// @route   POST /api/profile
// @desc    Create or Update user profile
// @access  Protected
router.post('/', passport.authenticate('jwt', { session: false }), (req, res) => {
    require('../../validation/profile')({ ...req.body, user: req.user.id })
        .then(profile => {
            Profile.findOneAndUpdate(
                { user: req.body.id },
                { $set: profile }, // form fields left empty must not be sent
                { new: true, upsert: true })
                .then(profile => res.json(profile))
                .catch(err => res.status(400).json(err))
        })
        .catch(errors => res.status(400).json(errors))
})


module.exports = router;