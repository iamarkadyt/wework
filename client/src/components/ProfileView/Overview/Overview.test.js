import React from 'react'
import { shallow } from 'enzyme'
import Overview from './Overview'
import EditButton from './EditButton/EditButton'
import FollowButton from './FollowButton/FollowButton'
import UnfollowButton from './UnfollowButton/UnfollowButton'
import cloneDeep from 'lodash.clonedeep'
import { mockProfile as profile } from '../../../mocks/profile'

let mountedComponent, props

const getMockProps = () => {
  return cloneDeep({
    profile,
    profileBelongsToAuthedUser: true,
    isFollowingProfileOwner: false,
    baseUrl: '/profile',
    history: { push: jest.fn() },
    followAPerson: jest.fn(),
    unfollowAPerson: jest.fn(),
    quitEntryDeletingMode: jest.fn(),
    navTo: jest.fn()
  })
}

const comp = () => {
  if (!mountedComponent) {
    mountedComponent = shallow(<Overview {...props} />)
  }
  return mountedComponent
}

describe('Overview', () => {
  beforeEach(() => {
    props = getMockProps()
    mountedComponent = undefined
  })

  describe('props', () => {
    describe('effects', () => {
      describe('profileBelongsToAuthedUser', () => {
        it('renders EditButton if true', () => {
          props.profileBelongsToAuthedUser = true
          expect(comp().find('.info-buttons').find(EditButton).exists()).toBe(true)
        })

        it('doesn\'t render EditButton if false', () => {
          props.profileBelongsToAuthedUser = false
          expect(comp().find('.info-buttons').find(EditButton).exists()).toBe(false)
        })
      })

      describe('isFollowingProfileOwner', () => {
        it('doesn\'t render UnfollowButton if false', () => {
          props.profileBelongsToAuthedUser = false
          props.isFollowingProfileOwner = false
          expect(comp().find('.info-buttons').find(UnfollowButton).exists()).toBe(false)
        })

        it('renders FollowButton if false', () => {
          props.profileBelongsToAuthedUser = false
          props.isFollowingProfileOwner = false
          expect(comp().find('.info-buttons').find(FollowButton).exists()).toBe(true)
        })

        it('doesn\'t render FollowButton if true', () => {
          props.profileBelongsToAuthedUser = false
          props.isFollowingProfileOwner = true
          expect(comp().find('.info-buttons').find(FollowButton).exists()).toBe(false)
        })

        it('renders UnfollowButton if true', () => {
          props.profileBelongsToAuthedUser = false
          props.isFollowingProfileOwner = true
          expect(comp().find('.info-buttons').find(UnfollowButton).exists()).toBe(true)
        })
      })
    })
  })

  describe('functions', () => {
    describe('navTo', () => {
      it('calls history.push', () => {
        
      })
    })
  })

  it('matches snapshot', () => {
    expect(comp()).toMatchSnapshot()
  })
})
