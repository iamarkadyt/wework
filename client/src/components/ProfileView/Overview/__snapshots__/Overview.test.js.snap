// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview matches snapshot 1`] = `
<section
  className="intro"
>
  <img
    alt=""
    src="gravatar://some.avatar"
  />
  <div
    className="info"
  >
    <name
      profile={
        Object {
          "__v": 0,
          "_id": "5b7b42ba57a5806b896e8f0d",
          "bio": "TEST BIO TEXT.",
          "company": "Some Company",
          "date": "2018-08-23T20:25:12.263Z",
          "education": Array [
            Object {
              "_id": "5b7cab633345cc4c066c0540",
              "current": true,
              "degree": "Nano CS Degree",
              "fieldOfStudy": "Bioengineering",
              "from": "2015-08-02T00:00:00.000Z",
              "school": "Some Bootcamp",
            },
          ],
          "experience": Array [
            Object {
              "_id": "5b7ca6cf63353349cf03756c",
              "company": "Some Company",
              "current": false,
              "from": "2009-03-31T23:00:00.000Z",
              "title": "Software Developer",
              "to": "2012-08-01T00:00:00.000Z",
            },
          ],
          "githubusername": "arkadyt",
          "handle": "john_doe",
          "location": "Seattle, WA",
          "skills": Array [
            "Go",
            "C#",
            "C++",
            "Whatnot else",
          ],
          "status": "Not available for employment",
          "title": "Senior Developer",
          "user": Object {
            "_id": "5b75f023f1957052e52106f5",
            "avatar": "gravatar://some.avatar",
            "name": "John Doe",
          },
          "website": "arkadyt.com",
        }
      }
    />
    <location
      profile={
        Object {
          "__v": 0,
          "_id": "5b7b42ba57a5806b896e8f0d",
          "bio": "TEST BIO TEXT.",
          "company": "Some Company",
          "date": "2018-08-23T20:25:12.263Z",
          "education": Array [
            Object {
              "_id": "5b7cab633345cc4c066c0540",
              "current": true,
              "degree": "Nano CS Degree",
              "fieldOfStudy": "Bioengineering",
              "from": "2015-08-02T00:00:00.000Z",
              "school": "Some Bootcamp",
            },
          ],
          "experience": Array [
            Object {
              "_id": "5b7ca6cf63353349cf03756c",
              "company": "Some Company",
              "current": false,
              "from": "2009-03-31T23:00:00.000Z",
              "title": "Software Developer",
              "to": "2012-08-01T00:00:00.000Z",
            },
          ],
          "githubusername": "arkadyt",
          "handle": "john_doe",
          "location": "Seattle, WA",
          "skills": Array [
            "Go",
            "C#",
            "C++",
            "Whatnot else",
          ],
          "status": "Not available for employment",
          "title": "Senior Developer",
          "user": Object {
            "_id": "5b75f023f1957052e52106f5",
            "avatar": "gravatar://some.avatar",
            "name": "John Doe",
          },
          "website": "arkadyt.com",
        }
      }
    />
    <p>
      Not available for employment
    </p>
    <links
      profile={
        Object {
          "__v": 0,
          "_id": "5b7b42ba57a5806b896e8f0d",
          "bio": "TEST BIO TEXT.",
          "company": "Some Company",
          "date": "2018-08-23T20:25:12.263Z",
          "education": Array [
            Object {
              "_id": "5b7cab633345cc4c066c0540",
              "current": true,
              "degree": "Nano CS Degree",
              "fieldOfStudy": "Bioengineering",
              "from": "2015-08-02T00:00:00.000Z",
              "school": "Some Bootcamp",
            },
          ],
          "experience": Array [
            Object {
              "_id": "5b7ca6cf63353349cf03756c",
              "company": "Some Company",
              "current": false,
              "from": "2009-03-31T23:00:00.000Z",
              "title": "Software Developer",
              "to": "2012-08-01T00:00:00.000Z",
            },
          ],
          "githubusername": "arkadyt",
          "handle": "john_doe",
          "location": "Seattle, WA",
          "skills": Array [
            "Go",
            "C#",
            "C++",
            "Whatnot else",
          ],
          "status": "Not available for employment",
          "title": "Senior Developer",
          "user": Object {
            "_id": "5b75f023f1957052e52106f5",
            "avatar": "gravatar://some.avatar",
            "name": "John Doe",
          },
          "website": "arkadyt.com",
        }
      }
    />
  </div>
  <div
    className="info-buttons"
  >
    <editButton
      navTo={[MockFunction]}
      quitEntryDeletingMode={[MockFunction]}
    />
  </div>
</section>
`;
